### ISSUE TITLE
hhvm/hphp/runtime/vm/native.cpp should really be using libffi

### ISSUE THREAD

**Issue by apinski-cavium** [facebook/hhvm#5035]
When running on ARM64, we found the code in hhvm/hphp/runtime/vm/native.cpp is not correct for returning structures.  It uses the first argument as the invisible argument but that does not follow the ABI which says this:
Otherwise, the caller shall reserve a block of memory of sufficient size and alignment to hold the result. The 
address of the memory block shall be passed as an additional argument to the function in x8. The callee may 
modify the result memory block at any point during the execution of the subroutine (there is no requirement for 
the callee to preserve the value stored in x8).
--- CUT ---

Instead of reinventing the wheel for each target, native.cpp should be converted into using libffi instead.

More information on libffi can be found at https://sourceware.org/libffi/ .  It is used by many different software already and has been ported to many different targets too including ARM64 and x86_64 and it even supports closures.

**Comment by apinski-cavium** [comment]
Here is the backtrace of the issue running on version.php in quick tests:
(gdb) r -vEval.Jit=0 -f ../test/quick/version.php
Starting program: /data/hhvm-generic/hhvm/hphp/hhvm/hhvm -vEval.Jit=0 -f ../test/quick/version.php
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
bool(true)
bool(true)
bool(true)

Program received signal SIGSEGV, Segmentation fault.
0x0000000001780a90 in HPHP::string_memnstr (haystack=<optimized out>, 
    needle=0x7fffffe5b0 "\300\345\377\377\177", needle_len=-1391460416, 
    end=0x8005b07659 <error: Cannot access memory at address 0x8005b07659>)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/zend-string.cpp:465
465        if ((p = (char *)memchr(p, *needle, (end-p+1))) && ne == p[needle_len-1]) {
(gdb) bt
#0  0x0000000001780a90 in HPHP::string_memnstr (haystack=<optimized out>, 
    needle=0x7fffffe5b0 "\300\345\377\377\177", needle_len=-1391460416, 
    end=0x8005b07659 <error: Cannot access memory at address 0x8005b07659>)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/zend-string.cpp:465
#1  0x0000000001780b90 in HPHP::string_find (input=0x7fb2c07618 ".", 
    len=<optimized out>, s=<optimized out>, s_len=<optimized out>, 
    pos=<optimized out>, case_sensitive=<optimized out>)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/zend-string.cpp:421
#2  0x0000000001775a2c in HPHP::String::find (this=this@entry=0x7fad0fff90, s=..., 
    pos=pos@entry=0, caseSensitive=caseSensitive@entry=true)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/type-string.cpp:189
#3  0x00000000017bec64 in HPHP::StringUtil::Explode (input=..., delimiter=..., 
    limit=-1391460480)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/string-util.cpp:86
#4  0x00000000023c7970 in HPHP::f_explode (delimiter=..., str=..., 
    limit=<optimized out>)
    at /data/hhvm-generic/hhvm/hphp/runtime/ext/string/ext_string.cpp:516
#5  0x000000000190ebd4 in HPHP::Native::callFuncInt64Impl (
    f=f@entry=0x23c7950 <HPHP::f_explode(HPHP::String const&, HPHP::String const&, int)>, GP=GP@entry=0x7fffffe400, GP_count=<optimized out>, 
    SIMD=SIMD@entry=0x7fffffe3c0, SIMD_count=<optimized out>)
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/native-func-caller.h:832
#6  0x000000000191a7d0 in HPHP::Native::callFunc<true, false> (
    func=func@entry=0x7fb2fd3560, ctx=ctx@entry=0x0, args=args@entry=0x7fad0fff90, 
    ret=...) at /data/hhvm-generic/hhvm/hphp/runtime/vm/native.cpp:176
#7  0x0000000001973000 in HPHP::iopFCallBuiltin (
    pc=@0x7fffffe598: 0x7fad6d785c "\r{")
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:6249
#8  0x0000000001a0bc50 in HPHP::dispatchImpl<false> ()
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:7588
#9  0x0000000001a13c04 in HPHP::dispatch ()
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:7594
#10 0x0000000001a1400c in HPHP::enterVMAtFunc (
    enterFnAr=enterFnAr@entry=0x7fad0fffc0, 
    stk=stk@entry=HPHP::StackArgsState::Untrimmed)
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:1974
#11 0x0000000001a14948 in HPHP::enterVM (ar=ar@entry=0x7fad0fffc0, 
    stk=stk@entry=HPHP::StackArgsState::Untrimmed, resumable=resumable@entry=0x0, 
    exception=exception@entry=0x0)
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:2031
#12 0x0000000001a2e8b0 in HPHP::ExecutionContext::invokeFunc (
---Type <return> to continue, or q <return> to quit---
    this=this@entry=0x7fb18064b0, retptr=retptr@entry=0x7fffffe840, 
    f=<optimized out>, args_=..., this_=this_@entry=0x0, cls=cls@entry=0x0, 
    varEnv=0x7fb1811240, invName=invName@entry=0x0, 
    flags=flags@entry=HPHP::ExecutionContext::InvokePseudoMain)
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:2198
#13 0x0000000001a2f244 in HPHP::ExecutionContext::invokeUnit (this=0x7fb18064b0, 
    retval=retval@entry=0x7fffffe840, unit=0x7facc07560)
    at /data/hhvm-generic/hhvm/hphp/runtime/vm/bytecode.cpp:2392
#14 0x000000000162e390 in HPHP::invoke_file_impl (res=..., path=..., 
    once=once@entry=true, currentDir=currentDir@entry=0x2def848 "")
    at /data/hhvm-generic/hhvm/hphp/runtime/base/builtin-functions.cpp:703
#15 0x0000000001631c98 in HPHP::invoke_file (s=..., once=once@entry=true, 
    currentDir=currentDir@entry=0x2def848 "")
    at /data/hhvm-generic/hhvm/hphp/runtime/base/builtin-functions.cpp:716
#16 0x0000000001631edc in HPHP::include_impl_invoke (file=..., 
    once=once@entry=true, currentDir=currentDir@entry=0x2def848 "")
    at /data/hhvm-generic/hhvm/hphp/runtime/base/builtin-functions.cpp:738
#17 0x0000000001795850 in HPHP::hphp_invoke (context=context@entry=0x7fb18064b0, 
    cmd=..., func=func@entry=false, funcParams=..., funcRet=..., reqInitFunc=..., 
    reqInitDoc=..., error=@0x7fffffea07: false, errorMsg=..., once=once@entry=true, 
    warmupOnly=warmupOnly@entry=false, richErrorMsg=richErrorMsg@entry=false)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/program-functions.cpp:1933
#18 0x0000000001795ae4 in HPHP::hphp_invoke_simple (filename=..., 
    warmupOnly=warmupOnly@entry=false)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/program-functions.cpp:1894
#19 0x000000000179eedc in HPHP::execute_program_impl (argc=argc@entry=4, 
    argv=argv@entry=0x7ffffff5a8)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/program-functions.cpp:1625
#20 0x000000000179faec in HPHP::execute_program (argc=argc@entry=4, 
    argv=argv@entry=0x7ffffff5a8)
    at /data/hhvm-generic/hhvm/hphp/runtime/base/program-functions.cpp:1025
#21 0x0000000001594eb0 in main (argc=4, argv=0x7ffffff5a8)
    at /data/hhvm-generic/hhvm/hphp/hhvm/main.cpp:60

**Comment by apinski-cavium** [comment]
That is if we look at:
#4 0x00000000023c7970 in HPHP::f_explode (delimiter=..., str=..., 
limit=)
at /data/hhvm-generic/hhvm/hphp/runtime/ext/string/ext_string.cpp:516

delimiter was garbage and str was what the delimiter should have been.

**Comment by paulbiss** [comment]
HHVM has a lot of architecture specific code, and we have to deal with calling conventions in multiple places including our back-end JIT compiler, so I'm not sure sucking in an entire library to do native calling would have many advantages.

**Comment by apinski-cavium** [comment]
Yes but it might but if you notice other JITs (OpenJDK, Dalvik and now even gccgo) use libffi for their native calling functions too.  It is would make the native function calling much simpler and is actually smaller than the code that is included in native.cpp/native-func-caller.h right now.  Plus you get closures for free.

**Comment by apinski-cavium** [comment]
Also there is no way to pass the return value in C code so the code as implemented will never work on AARCH64.  So either writing native calling code in assembly is needed.  Why reinvent the code here and not use libffi as it is already done correctly in libffi?

**Comment by edwinsmith** [comment]
How hard would it be to make the libffi dependency only on ARM64?
Could you post a pull request with the libffi changes so we can see how intrusive they are?

Paul, other than "one more dependency" is there anything about libffi that makes it something we don't want to include? Because the ARM abi requires register x8 to hold the implicit struct pointer, there's no mangled-c-style call to make this work (unlike x64 which uses the first argument register). I'm also wondering if the PPC64 project is going to need basically the same thing.

**Comment by edwinsmith** [comment]
@oyamauchi did you run into this on earlier arm64 hardware? i can easily imagine this not being a problem on vixl, but not sure.

**Comment by paulbiss** [comment]
@edwinsmith: The licensing looks fine to me, assuming it can be turned off for x64, and it doesn't create a major maintenance burden I have no objection to adding it.

**Comment by oyamauchi** [comment]
I didn't run into this. I don't know if there even were any C++ helpers that returned structs when I originally ran on ARM hardware. It might have been before we started passing around Cells and TypedValues by value.

**Comment by edwinsmith** [comment]
@apinski-cavium sounds like FFI will be fine, if you can organize the code so it's only used for ARM. @oyamauchi i think this is for when the interpreter needs to call a builtin function, so the typical case is a Variant return value, or one of the other pointer wrappers.

**Comment by ykarlsbrun** [comment]
The code below (native.cpp) does not work for arm64 because return value is put in GP_args[0].

template<bool usesDoubles, bool variadic>
void callFunc(const Func* func, void *ctx,
              TypedValue *args, TypedValue& ret) {
  assert(variadic == func->hasVariadicCaptureParam());

  int64_t GP_args[kMaxBuiltinArgs];
  double SIMD_args[kNumSIMDRegs];
  int GP_count = 0, SIMD_count = 0;

  auto const numArgs = func->numParams();
  auto retType = func->returnType();

  if (!retType) {
    GP_args[GP_count++] = (int64_t)&ret;
  } else if (isBuiltinByRef(retType)) {
    GP_args[GP_count++] = (int64_t)&ret.m_data;
  }

**Comment by aorenste** [comment]
this has been fixed
